File: npm/kirkforge-plugin/packages/memory-palace/src/adapters/in-memory.ts
import { ok, type Result } from "@kirkforge/core-types";
import type {
  MemoryAdapter,
  MemoryObject,
  MemoryQuery,
  MemoryStats,
} from "../types.js";

export class InMemoryAdapter implements MemoryAdapter {
  private objects: MemoryObject[] = [];

  async write(obj: MemoryObject): Promise<Result<void, Error>> {
    this.objects.push(obj);
    return ok(undefined);
  }
  async read(id: string): Promise<Result<MemoryObject | null, Error>> {
    return ok(this.objects.find((o) => o.id === id) ?? null);
  }
  async query(q: MemoryQuery): Promise<Result<MemoryObject[], Error>> {
    let results = [...this.objects];
    if (q.kind) results = results.filter((o) => o.kind === q.kind);
    if (q.tags) results = results.filter((o) => q.tags!.some((t) => o.tags.includes(t)));
    if (q.since) results = results.filter((o) => o.timestamp >= q.since!);
    results.sort((a, b) => b.timestamp.localeCompare(a.timestamp));
    if (q.limit) results = results.slice(0, q.limit);
    return ok(results);
  }
  async stats(): Promise<Result<MemoryStats, Error>> {
    return ok({
      totalObjects: this.objects.length,
      lastWrite: this.objects[this.objects.length - 1]?.timestamp ?? "never",
    });
  }
  async persist(): Promise<void> {
    /* no-op: in-memory only */
  }
  clear(): void {
    this.objects = [];
  }
}

File: crates/kf-memory-store/src/adapters/in_memory.rs
//! In-memory adapter. Port of `memory-palace/src/adapters/in-memory.ts`.

use std::sync::Mutex;

use anyhow::Result;

use crate::adapters::MemoryAdapter;
use crate::types::{MemoryObject, MemoryQuery, MemoryStats};

pub struct InMemoryAdapter {
    objects: Mutex<Vec<MemoryObject>>,
}

impl Default for InMemoryAdapter {
    fn default() -> Self {
        Self::new()
    }
}

impl InMemoryAdapter {
    pub fn new() -> Self {
        Self {
            objects: Mutex::new(Vec::new()),
        }
    }

    pub fn clear(&self) {
        self.objects
            .lock()
            .expect("in-memory lock poisoned")
            .clear();
    }
}

impl MemoryAdapter for InMemoryAdapter {
    fn write(&self, obj: &MemoryObject) -> Result<()> {
        self.objects
            .lock()
            .expect("in-memory lock poisoned")
            .push(obj.clone());
        Ok(())
    }

    fn read(&self, id: &str) -> Result<Option<MemoryObject>> {
        let guard = self.objects.lock().expect("in-memory lock poisoned");
        Ok(guard.iter().find(|o| o.id == id).cloned())
    }

    fn query(&self, q: &MemoryQuery) -> Result<Vec<MemoryObject>> {
        let guard = self.objects.lock().expect("in-memory lock poisoned");
        let mut results: Vec<MemoryObject> = guard
            .iter()
            .filter(|o| q.kind.as_deref().is_none_or(|k| o.kind == k))
            .filter(|o| {
                q.tags
                    .as_deref()
                    .is_none_or(|tags| tags.iter().any(|t| o.tags.iter().any(|ot| ot == t)))
            })
            .filter(|o| q.since.as_deref().is_none_or(|s| o.timestamp.as_str() >= s))
            .cloned()
            .collect();
        results.sort_by(|a, b| b.timestamp.cmp(&a.timestamp));
        if let Some(limit) = q.limit {
            results.truncate(limit);
        }
        Ok(results)
    }

    fn stats(&self) -> Result<MemoryStats> {
        let guard = self.objects.lock().expect("in-memory lock poisoned");
        let last_write = guard
            .last()
            .map(|o| o.timestamp.clone())
            .unwrap_or_else(|| "never".to_string());
        Ok(MemoryStats {
            total_objects: guard.len(),
            last_write,
        })
    }
}
